### PR TITLE
[SPARK-36074][SQL] Add error class for StructType.findNestedField

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR adds an INVALID_FIELD_NAME error class for the errors in `StructType.findNestedField`. It also cleans up the code there and adds UT for this method.

### Why are the changes needed?

follow the new error message framework

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @karenfeng @HyukjinKwon

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -21,5 +21,8 @@
   "WRITING_JOB_ABORTED" : {
     "message" : [ "Writing job aborted" ],
     "sqlState" : "40000"
+  },
+  "INVALID_FIELD_NAME" : {
```
Comment (yaooqinn) [inline core/src/main/resources/error/error-classes.json:25]: can we sort alphabetically by the keys

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -1349,9 +1349,13 @@ private[spark] object QueryCompilationErrors {
         s"${evalTypes.mkString(",")}")
   }
 
-  def ambiguousFieldNameError(fieldName: String, names: String): Throwable = {
+  def ambiguousFieldNameError(
+      fieldName: Seq[String], nunMatches: Int, context: Origin): Throwable = {
     new AnalysisException(
-      s"Ambiguous field name: $fieldName. Found multiple columns that can match: $names")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1354]: Is it
```
46121 | X | 46 | OLB-specific error | 121 | invalid column name
```
?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -1349,9 +1349,13 @@ private[spark] object QueryCompilationErrors {
         s"${evalTypes.mkString(",")}")
   }
 
-  def ambiguousFieldNameError(fieldName: String, names: String): Throwable = {
+  def ambiguousFieldNameError(
+      fieldName: Seq[String], nunMatches: Int, context: Origin): Throwable = {
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1353]: Nit: `nunMatches` -> `numMatches`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -1349,9 +1349,13 @@ private[spark] object QueryCompilationErrors {
...
     new AnalysisException(
-      s"Ambiguous field name: $fieldName. Found multiple columns that can match: $names")
+      errorClass = "INVALID_FIELD_NAME",
+      messageParameters = Array(fieldName.quoted,
+        s"${fieldName.last} is ambiguous and has $nunMatches matching fields in the struct"),
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1357]: Can we put the majority of this in the error message format? Eg. the format would be `Field name %s is invalid: %s is ambiguous and has %s matching fields in the struct`, and the parameters would be `Array(fieldName.quoted, fieldName.last, numMatches)`

Comment (karenfeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1357]: In this case, we would have to create two new error classes (another one for the case that path is not a struct).

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -7,6 +7,9 @@
     "message" : [ "Found duplicate keys '%s'" ],
     "sqlState" : "23000"
   },
+  "INVALID_FIELD_NAME" : {
+    "message" : [ "Field name %s is invalid: %s." ]
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:11]: I believe this is actually SQLSTATE 42000 (syntax error) as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:1354]: I don't know if this is `OLB-specific error`. I'll pick `SQLSTATE 42000 (syntax error)` as @karenfeng suggested

**Review by karenfeng — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/3.2 (since error class is a new feature in 3.2)

**Pull request closed without merging**